Sources/TensorFlow/Layers/Normalization.swift

```diff
@@ -224,3 +244,166 @@ public struct LayerNorm<Scalar: TensorFlowFloatingPoint>: Layer {
+  ///   - featureCount: The number of features.
+  ///   - groups: The number of groups.
+  ///   - axis: The axis that should be normalized.
+  ///   - epsilon: The small scalar added to variance.
+  public init(
+    featureCount: Int,
+    groups: Int,
+    axis: Int = -1,
+    epsilon: Scalar = 1e-3
+  ) {
+    precondition(
+      featureCount.isMultiple(of: groups),
+      "The feature count must be divisible by groups.")
+    self.init(
+      offset: Tensor(zeros: [featureCount]),
+      scale: Tensor(ones: [featureCount]),
+      groups: groups,
+      axis: axis,
+      epsilon: epsilon
+    )
+  }
+
+  /// Returns the output obtained from applying the layer to the given input.
+  ///
+  /// - Parameter input: The input to the layer.
+  /// - Returns: The output.
+  @differentiable
+  public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+    let positiveAxis = (input.rank + axis) % input.rank
+    precondition(positiveAxis != 0, "The axis cannot be batch axis.")
+    var offset = self.offset
+    var scale = self.scale
+    var broadcastShape = TensorShape([Int](repeating: 1, count: input.rank + 1))
+    broadcastShape[positiveAxis] = groups
+    broadcastShape[positiveAxis + 1] = input.shape[positiveAxis] / groups
+    offset = offset.reshaped(to: broadcastShape)
+    scale = scale.reshaped(to: broadcastShape)
+
+    var groupShape = input.shape
+    groupShape[positiveAxis] /= groups
+    groupShape.insert(groups, at: positiveAxis)
+    let grouped = input.reshaped(to: groupShape)
+    var normalizedAxes = Array(1..<grouped.rank)
+    normalizedAxes.remove(at: positiveAxis - 1)
+    let moments = grouped.moments(alongAxes: normalizedAxes)
+    let eps = Tensor(epsilon)
+    let normalized = _batchNorm(
+      grouped,
+      mean: moments.mean, variance: moments.variance,
+      offset: offset, scale: scale,
+      varianceEpsilon: eps)
+    return normalized.reshaped(to: input.shape)
+  }
+}
+
+/// A layer that applies instance normalization over a mini-batch of inputs.
+///
+/// Reference: [Instance Normalization: The Missing Ingredient for Fast Stylization](https://arxiv.org/abs/1607.08022).
+@frozen
+public struct InstanceNorm<Scalar: TensorFlowFloatingPoint>: Layer {
+  /// Internal group normalization.
+  var groupNorm: GroupNorm<Scalar>
+
+  /// The offset value, also known as beta.
+  public var offset: Tensor<Scalar> {
+    _read { yield groupNorm.offset }
+    _modify { yield &groupNorm.offset }
+  }
+
+  /// The scale value, also known as gamma.
+  public var scale: Tensor<Scalar> {
+    _read { yield groupNorm.scale }
+    _modify { yield &groupNorm.scale }
+  }
+
+  /// The axis.
+  public var axis: Int { groupNorm.axis }
+  /// The variance epsilon value.
+  public var epsilon: Scalar { groupNorm.epsilon }
+
+  /// Creates a instance normalization layer.
+  public init(
+    offset: Tensor<Scalar>,
+    scale: Tensor<Scalar>,
+    axis: Int,
+    epsilon: Scalar
+  ) {
+    groupNorm = GroupNorm(
```

Comment:
The use of `GroupNorm` as an implementation detail implies all the `init`s of`InstaneNorm` inherit certain preconditions that need to be documented… otherwise it's not an implementation detail! :wink: